[16.0][MIG] pos_customer_wallet_partner_is_user

carmenbianca opened the pull request:
## Description

Depends on #300 

## Odoo task (if applicable)

https://gestion.coopiteasy.be/web#id=10798&action=475&active_id=492&model=project.task&view_type=form&menu_id=536

## Checklist before approval

- [ ] Tests are present (or not needed).
- [ ] Credits/copyright have been changed correctly.
- [ ] Change log snippet is present.
- [ ] (If a new module) Moving this to OCA has been considered.

carmenbianca committed:
[ADD] pos_customer_wallet_is_user

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REM] pos_customer_wallet_partner_is_user: Don't compute the field

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] SCRLfs → SC

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

Commit:
[UPD] Update pos_customer_wallet_partner_is_user.pot

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
[UPD] README.rst

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
pos_customer_wallet_partner_is_user 12.0.1.1.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] pos_customer_wallet[_*]: Rename foodprint class

The class name was wrongly copied from some other place and does not
reflect what it's trying to represent.

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] pos_customer_wallet_partner_is_user: Raise error when functionality disabled

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] pos_customer_wallet_partner_is_user: Add to roadmap

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] pos_customer_wallet_partner_is_user: Use more robust method to detect wallet usage

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] pos_customer_wallet[_partner_is_user]: Pass required argument

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] pos_customer_wallet_partner_is_user: Clarifying comment

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

Commit:
[UPD] Update pos_customer_wallet_partner_is_user.pot

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
[UPD] README.rst

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
pos_customer_wallet_partner_is_user 12.0.1.2.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] pos_customer_wallet_[*]: Repair maintainer tags in manifest

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[UPD] Run prettier

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
[UPD] README.rst

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
pos_customer_wallet_partner_is_user 12.0.1.2.1

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

Commit:
[UPD] Update pos_customer_wallet_partner_is_user.pot

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] pos_customer_wallet_partner_is_user: black, isort, prettier

carmenbianca committed:
[MIG] pos_customer_wallet_partner_is_user: Migration to 13.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[MIG] pos_customer_wallet_partner_is_user: Migration to 14.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[MIG] pos_customer_wallet_partner_is_user: Migration to 15.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[MIG] pos_customer_wallet_partner_is_user: Migration to 16.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

polchampion reviewed (changes requested):
Functionally tested 22/01/2024 on [16-test-portal-customer-wallet](https://odoo16.main.test.srv.coopiteasy.be/web?db=16-test-portal-customer-wallet)

@carmenbianca See issues below + the feature list was imporved as follows.

- [ ] Adds a boolean field on partners that shows whether they have used the customer wallet functionality **--> Looks like the field isn't computed based on wallet credit or usage; also it can be changed manually. Unclear to me whether customer_partner_is_user needs fixing. I read elsewhere that this bool's intention was to display the balance on the pos ticket, in which case I was mistaken and the module's aim isn't to compute whether the partner is a customer wallet user, but instead to make it possible for a user to flag partners as wallet users - this porbably needs clarfying in the readme**
- [x] Adds two search filters: is_customer_wallet_user and a filter on non-zero customer wallet balance
- [ ] forbids certain accounting operations from happening if the partner is not a customer wallet user **--> what operations ?**
- [ ] Hide the customer wallet payment method if the selected partner is not a user. **--> NOK**
- [ ] Hide the customer wallet balance if the selected partner is not a user. ** don't know since the balance is never displayed yet**

polchampion commented:
@carmenbianca I updated the above review just now

carmenbianca committed:
[IMP] pos_customer_wallet_partner_is_user: Hide payment method conditionally

If partner is not a user, don't show the payment method.

The deleted JS code is made obsolete/superfluous because the payment
method cannot even be selected now.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented:
> Looks like the field isn't computed based on wallet credit or usage; also it can be changed manually.

It was always a manual field. This has not changed in the migration.

>  forbids certain accounting operations from happening if the partner is not a customer wallet user --> what operations ?

It forbids the user from using the wallet payment method in the POS. An error shows up. However, I may delete this code because…

>  Hide the customer wallet payment method if the selected partner is not a user. --> NOK

I just added this. This makes the above code redundant.

>  Hide the customer wallet balance if the selected partner is not a user. ** don't know since the balance is never displayed yet**

A few notes:

- The balance _is_ displayed. I added the customer wallet payment method to the POS in the test server. 'Customer Account' is something else entirely that is confusingly similarly named, so I've removed that payment method.
- Because of the above change, the balance is never shown for non-users in the POS UI.
- However, specifically, the customer wallet balance is invisible on the _receipts_ of non-users.

---

I updated the code on the test server.

polchampion reviewed (approved):
Sucessfull functional test

@carmenbianca Thanks for your explanations, lgtm. Feel free to include anything missing in the readme from the updated features list that follows.

- [x] Adds an editable boolean field to partners (`is_customer_wallet_user`) intended to flag users of the customer wallet
- [x]  Adds two search filters: is_customer_wallet_user and a filter on non-zero customer wallet balance
- [x] Hides the customer wallet payment method on the POS payment UI if the selected client is not a "customer wallet user"
- [x] Only displays the customer wallet balance on the POS ticket if the client is a "customer wallet user"
